README.rst
errorguimonitor
====================================================

command line utility to send notifications about new/increased errors from ErrorGUI
-------------------------------------------

.. image:: https://secure.travis-ci.org/imalkevich/errorguimonitor.png?branch=master
        :target: https://travis-ci.org/imalkevich/errorguimonitor

.. image:: https://codecov.io/github/imalkevich/errorguimonitor/coverage.svg?branch=master
    :target: https://codecov.io/github/imalkevich/errorguimonitor
    :alt: codecov.io

Are you watching for module and would like to understand whether new code being added is 
not causing new issues - run this tool on a regular basis to be able to see the trends in errors.

Installation
------------

::

    pip install errorguimonitor

or

::

    python setup.py install

Usage
-----

::

    usage: reporter.py [-h] [-e ENVIRONMENT] [-d DATE] [-u USER] [-p PASSWORD]
                   [-smtp_user SMTP_USER] [-smtp_password SMTP_PASSWORD] [-v]

    compare errors for two periods in s specific environment via the command line

    optional arguments:
    -h, --help            show this help message and exit
    -e ENVIRONMENT, --environment ENVIRONMENT
                            the environment (QED only at the moment)
    -d DATE, --date DATE  the target date (year-month-day)
    -u USER, --user USER  SAFE ID user
    -p PASSWORD, --password PASSWORD
                            SAFE ID password
    -smtp_user SMTP_USER, --smtp_user SMTP_USER
                            Gmail account username
    -smtp_password SMTP_PASSWORD, --smtp_password SMTP_PASSWORD
                            Gmail account password
    -v, --version         displays the current version of errorguimonitor

Author
------

-  Ihar Malkevich (imalkevich@gmail.com)

CHANGES.txt
0.1
------

-  I'am doing it live!

test_errorguimonitor.py
#!/usr/bin/env python

""" Tests for errorguimonitor. """

import inspect
import os
import re
import sys
import unittest
import unittest.mock as mock

from datetime import datetime, timedelta
from pyquery import PyQuery as pq

from errorguimonitor import reporter, report_builder, report_retriever

class ErrorCompareReporterTestCase(unittest.TestCase):
    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_get_report_dates_today_is_weekend(self):
        """ test today is Saturday, 6 days back """
        # arrange
        today = datetime(2018, 4, 7) + timedelta(hours=15)
        qed1week = reporter.ErrorCompareReporter('env', today, 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')
        
        # act
        target_dates, history_dates = qed1week._get_report_dates(1, 5)

        # assert
        self.assertEqual(len(target_dates), 1)
        self.assertEqual(target_dates[0], (datetime(2018, 4, 6), datetime(2018, 4, 6, 23, 59, 59)))

        self.assertEqual(len(history_dates), 5)
        self.assertEqual(history_dates[0], (datetime(2018, 4, 5), datetime(2018, 4, 5, 23, 59, 59)))
        self.assertEqual(history_dates[1], (datetime(2018, 4, 4), datetime(2018, 4, 4, 23, 59, 59)))
        self.assertEqual(history_dates[2], (datetime(2018, 4, 3), datetime(2018, 4, 3, 23, 59, 59)))
        self.assertEqual(history_dates[3], (datetime(2018, 4, 2), datetime(2018, 4, 2, 23, 59, 59)))
        self.assertEqual(history_dates[4], (datetime(2018, 3, 30), datetime(2018, 3, 30, 23, 59, 59)))

    def test_get_report_dates(self):
        """ test today is Tuesday, 6 days back """
        # arrange
        today = datetime(2018, 4, 10) + timedelta(hours=8)
        qed1week = reporter.ErrorCompareReporter('env', today, 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')
        
        # act
        target_dates, history_dates = qed1week._get_report_dates(1, 5)

        # assert
        self.assertEqual(len(target_dates), 1)
        self.assertEqual(target_dates[0], (datetime(2018, 4, 10), datetime(2018, 4, 10, 23, 59, 59)))

        self.assertEqual(len(history_dates), 5)
        self.assertEqual(history_dates[0], (datetime(2018, 4, 9), datetime(2018, 4, 9, 23, 59, 59)))
        self.assertEqual(history_dates[1], (datetime(2018, 4, 6), datetime(2018, 4, 6, 23, 59, 59)))
        self.assertEqual(history_dates[2], (datetime(2018, 4, 5), datetime(2018, 4, 5, 23, 59, 59)))
        self.assertEqual(history_dates[3], (datetime(2018, 4, 4), datetime(2018, 4, 4, 23, 59, 59)))
        self.assertEqual(history_dates[4], (datetime(2018, 4, 3), datetime(2018, 4, 3, 23, 59, 59)))

    def test_get_report(self):
        """ Test get report merged from multiple files into one """
        # arrange
        vals = {
            (datetime(2018, 4, 5), datetime(2018, 4, 5, 23, 59, 59)): { 'module_1': { 'a': 1, 'c': 1 }, 'module_2': { 'x': 1, 'y': 2 } }, 
            (datetime(2018, 4, 4), datetime(2018, 4, 4, 23, 59, 59)): { 'module_1': { 'b': 2, 'c': 2 }, 'module_2': { 'x': 2 } }
        }

        def side_effect(*args):
            return vals[args]

        mock_report_builder = report_builder.ErrorReportBuilder('env', 'user', 'password')
        mock_report_builder.build_report = mock.MagicMock(side_effect=side_effect)

        qed1week = reporter.ErrorCompareReporter('env', datetime.today(), 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')
        dates = [
            (datetime(2018, 4, 5), datetime(2018, 4, 5, 23, 59, 59)),
            (datetime(2018, 4, 4), datetime(2018, 4, 4, 23, 59, 59))
        ]
        
        # act
        report = qed1week._get_report(mock_report_builder, dates)

        # assert
        self.assertEqual(report, { 'module_1': { 'a': [1], 'b': [2], 'c': [1, 2] }, 'module_2': { 'x': [1, 2], 'y': [2] } })

    def test_check_error_rate_increase(self):
        # arrange
        qed1week = reporter.ErrorCompareReporter('env', datetime.today(), 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')
        target_errors = [10]
        history_errors = [5]

        # act
        conf_int_diff = qed1week._check_error_rate_increase(target_errors, history_errors)

        # assert
        self.assertTrue(conf_int_diff[0] > 0)

    def test_prepare_report(self):
        # arrange
        qed1week = reporter.ErrorCompareReporter('env', datetime.today(), 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')
        target_report = { 
            'Document': { 'doc_key': [1] },
            'Website': { 'error_key_1': [1], 'error_key_2': [2], 'error_key_3': [3] }
        }

        history_report = { 
            'Document': { 'doc_key': [1] },
            'Website': { 'error_key_2': [2, 3, 4, 5], 'error_key_3': [3, 4] }
        }

        def side_effect(target_errors_count, history_errors_count):
            if target_errors_count == [2] \
                and history_errors_count == [2, 3, 4, 5]:
                return [1.0, 2.0]
            elif target_errors_count == [3] \
                and history_errors_count == [3, 4]:
                return [0.0, 0.5]
            
            raise ValueError('unexpected arguments')

        qed1week._check_error_rate_increase = mock.MagicMock(side_effect=side_effect)

        # act
        (new_errors, errors_increased) = qed1week._prepare_report(target_report, history_report)

        # assert
        self.assertEqual(new_errors, { 'error_key_1': [1] })
        self.assertEqual(errors_increased, { 'error_key_2': [1.0, 2.0] })

    def test_create_notification(self):
        # arrange
        qed1week = reporter.ErrorCompareReporter('env', datetime.today(), 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')

        target_dates = [
            (datetime(2018, 4, 9), datetime(2018, 4, 9, 23, 59, 59))
        ]

        history_dates = [
            (datetime(2018, 4, 6), datetime(2018, 4, 6, 23, 59, 59)),
            (datetime(2018, 4, 5), datetime(2018, 4, 5, 23, 59, 59))
        ]

        new_errors = { 'error_key_1': [1] }
        errors_increased = { 'error_key_2': [1.0, 2.0], 'error_key_3': [.5, .8] }

        # act
        (subject, body) = qed1week._create_notification(target_dates, history_dates, new_errors, errors_increased)

        # assert
        self.assertTrue(len(subject) > 0)
        self.assertTrue(len(body) > 0)

    @mock.patch('errorguimonitor.reporter.smtplib')
    def test_send_notification(self, mock_smtplib):
        # arrange
        qed1week = reporter.ErrorCompareReporter('env', datetime.today(), 'safeid_user', 'safeid_password', 'smtp_user', 'smtp_password')

        mock_server = mock.MagicMock()
        dummySMTP = mock.MagicMock(return_value=mock_server)

        mock_smtplib.SMTP = dummySMTP

        # act
        qed1week._send_notification('Test email', 'Test email body', 'smtp_user', 'smtp_password', ['ihar.malkevich@thomsonreuters.com'])

        # assert
        mock_smtplib.SMTP.assert_called_with('smtp.gmail.com', 587)
        mock_server.ehlo.assert_called_once()
        mock_server.starttls.assert_called_once()
        mock_server.login.assert_called_with('smtp_user', 'smtp_password')
        mock_server.send_message.assert_called_once()
        mock_server.quit.assert_called_once()

class ErrorReportBuilderTestCase(unittest.TestCase):
    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_load_report(self):
        # arrange
        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')

        builder.report_retriever.get_html = mock.MagicMock(return_value='<html></html>')
        builder._process_html = mock.MagicMock(return_value={ 'module': { 'key': 1 } })

        start_time = datetime(2018, 4, 8)
        end_time = datetime(2018, 4, 8, 23, 59, 59)

        # act
        report = builder._load_report(start_time, end_time)

        # assert
        self.assertEqual(report, { 'module': { 'key': 1 } })

        
        builder.report_retriever.get_html.assert_called_with(start_time, end_time)

    def test_process_html(self):
        # arrange
        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')

        html_path = os.path.join(
            os.path.dirname(inspect.getfile(self.__class__)),
            'test_fixture',
            'sample_errorgui_page.html')

        with open(html_path, 'r') as f:
            content = f.read()

        html = pq(content)

        # act
        report = builder._process_html(html)

        # assert
        self.assertEqual(
            report, 
            {
                'Document': { 'document_error_key_1': 1, 'document_error_key_2': 2 },
                'Search': { 'search_error_key': 2 },
                'Website': { 'website_error_key_1': 3, 'website_error_key_2': 5 }
            }
        )

    @mock.patch('errorguimonitor.report_builder.os')    
    def test_get_history_data_path_dir_not_exist(self, mock_os):
        # arrange
        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')
        mock_os.path.join.return_value = 'history_data'
        mock_os.path.isdir.return_value = False
        mock_os.makedirs.return_value = True

        # act
        history_dir = builder._get_history_data_path()

        # assert
        expected_dir = 'history_data'
        self.assertEqual(history_dir, expected_dir)

        mock_os.path.isdir.assert_called_with(expected_dir)
        mock_os.makedirs.assert_called_with(expected_dir)

    @mock.patch('errorguimonitor.report_builder.os')
    def test_get_history_data_path_dir_exist(self, mock_os):
        # arrange
        mock_os.path.join.return_value = 'history_data'
        mock_os.path.isdir.return_value = True
        mock_os.makedirs.return_value = False

        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')

        # act
        history_dir = builder._get_history_data_path()

        # assert
        expected_dir = 'history_data'
        self.assertEqual(history_dir, expected_dir)

        mock_os.path.isdir.assert_called_with(expected_dir)
        self.assertFalse(mock_os.makedirs.called)

    @mock.patch('errorguimonitor.report_builder.os')
    @mock.patch('errorguimonitor.report_builder.pickle')
    @mock.patch('errorguimonitor.report_builder.open')
    def test_build_report_no_report_exist(self, mock_open, mock_pickle, mock_os):
        # arrange
        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')

        builder._get_history_data_path = mock.MagicMock(return_value='history_data')
        builder._load_report = mock.MagicMock(return_value={ 'module': { 'key': 1 } })

        mock_os.path.join.return_value = 'path_does_not_exist'
        mock_os.path.isfile.return_value = False

        mock_pickle.dump.return_value = True

        mock_open.return_value = True

        start_time = datetime(2018, 4, 8)
        end_time = datetime(2018, 4, 8, 23, 59, 59)

        # act
        report = builder.build_report(start_time, end_time)

        # assert
        self.assertEqual(report, { 'module': { 'key': 1 } })

        mock_os.path.join.assert_called_with('history_data', '2018_04_08_00_00__2018_04_08_23_59.pkl')
        mock_os.path.isfile.assert_called_with('path_does_not_exist')
        builder._load_report.assert_called_with(start_time, end_time)
        mock_open.assert_called_with('path_does_not_exist', 'wb')
        mock_pickle.dump.assert_called_with(report, True)

    @mock.patch('errorguimonitor.report_builder.os')
    @mock.patch('errorguimonitor.report_builder.pickle')
    @mock.patch('errorguimonitor.report_builder.open')
    def test_build_report_report_exist(self, mock_open, mock_pickle, mock_os):
        # arrange
        builder = report_builder.ErrorReportBuilder('some_env', 'user', 'password')

        builder._get_history_data_path = mock.MagicMock(return_value='history_data')

        mock_os.path.join.return_value = 'path_exists'
        mock_os.path.isfile.return_value = True

        mock_pickle.load.return_value = { 'module': { 'key': 1 } }

        mock_open.return_value = True

        start_time = datetime(2018, 4, 8)
        end_time = datetime(2018, 4, 8, 23, 59, 59)

        # act
        report = builder.build_report(start_time, end_time)

        # assert
        self.assertEqual(report, { 'module': { 'key': 1 } })

        mock_os.path.join.assert_called_with('history_data', '2018_04_08_00_00__2018_04_08_23_59.pkl')
        mock_os.path.isfile.assert_called_with('path_exists')
        mock_open.assert_called_with('path_exists', 'rb')
        mock_pickle.load.assert_called_with(True)

class ErrorReportRetrieverTestCase(unittest.TestCase):
    def test_get_error_gui_url(self):
        # arrange
        start_time = datetime(2018, 4, 8)
        end_time = datetime(2018, 4, 8, 23, 59, 59)
        
        retriever = report_retriever.ErrorReportRetriever('QED', 'user', 'password')

        # act
        url = retriever._get_error_gui_url(start_time, end_time)

        # assert
        expected_url = (
            "http://cobalttools.qed.int.westgroup.com/ErrorGUINext/initialResults.do?search=y&param=&qS=&a=&b=&c=&d=&e=&f=&g=&h=&i=&j=&x=&y=&z=&v=&"
            "environs=QED&view=summary&site=all&start_time=2018-04-08%2000%3A00&end_time=2018-04-08%2023%3A59&submitForm=Submit"
        )
        self.assertEqual(url, expected_url)

    '''
    @mock.patch('errorguimonitor.report_retriever.requests')
    def test_get_html(self, mock_requests):
        # arrange
        start_time = datetime(2018, 4, 8)
        end_time = datetime(2018, 4, 8, 23, 59, 59)
        


        retriever = report_retriever.ErrorReportRetriever('QED', 'user', 'password')

        # act
        html = retriever.get_html(start_time, end_time)

        # assert
        self.assertIsNotNone(html)
    '''

if __name__ == '__main__':
    unittest.main()

errorguimonitor/report_builder.py
#!/usr/bin/env python

"""
Report builder module.
"""

import inspect
import os
import pickle

from urllib.parse import quote as url_quote

from pyquery import PyQuery as pq

from .report_retriever import ErrorReportRetriever

TIME_FORMAT = '%Y_%m_%d_%H_%M'

class ErrorReportBuilder(object):
    def __init__(self, env, user, password):
        self.report_retriever = ErrorReportRetriever(env, user, password)

    def build_report(self, start_time, end_time):
        report_file_name = os.path.join(
            self._get_history_data_path(), 
            '{}__{}.pkl'.format(start_time.strftime(TIME_FORMAT), end_time.strftime(TIME_FORMAT))
        )
        
        if not os.path.isfile(report_file_name):
            report = self._load_report(start_time, end_time)
            pickle.dump(report, open(report_file_name, 'wb'))
        else:
            report = pickle.load(open(report_file_name, 'rb'))

        return report

    def _extract_error(self, tr, error_dict):
        app = tr('td:eq(1)').text().strip()
        key = tr('td:eq(2)').text().strip()
        occurrences = int(tr('td:eq(3)').text().strip())

        if app not in error_dict:
            error_dict[app] = {}

        if key not in error_dict[app]:
            error_dict[app][key] = occurrences

        return error_dict

    def _get_history_data_path(self):
        history_dir = os.path.join(
            '../',
            os.path.dirname(inspect.getfile(self.__class__)),
            'history_data')

        if not os.path.isdir(history_dir):
             os.makedirs(history_dir)

        return history_dir

    def _load_report(self, start_time, end_time):
        result = self.report_retriever.get_html(start_time, end_time)
        html = pq(result)

        return self._process_html(html)

    def _process_html(self, html):
        errors = {}
        trs = html('div.borderedBoxWhite table.tablesorter tr')[1:]
        for tr in trs:
            self._extract_error(pq(tr), errors)

        return errors



errorguimonitor/report_retriever.py
#!/usr/bin/env python

"""
Report retriever module.

Grab HTML from the url provided.
"""

import random
import requests

from urllib.parse import quote as url_quote

from pyquery import PyQuery as pq

USER_AGENTS = ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:11.0) Gecko/20100101 Firefox/11.0',
               'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:22.0) Gecko/20100 101 Firefox/22.0',
               'Mozilla/5.0 (Windows NT 6.1; rv:11.0) Gecko/20100101 Firefox/11.0',
               ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_4) AppleWebKit/536.5 (KHTML, like Gecko) '
                'Chrome/19.0.1084.46 Safari/536.5'),
               ('Mozilla/5.0 (Windows; Windows NT 6.1) AppleWebKit/536.5 (KHTML, like Gecko) Chrome/19.0.1084.46'
                'Safari/536.5'), )

ENV = {
    'QED': {
        'safe_url': 'https://safe.thomson.com/safe-ui/fcc/login.fcc',
        'headers': {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.9'
        },
        'payload': (
            "action=LoginPassThrough&AcctMaintPurpose=&ProtectionLevel=SAFE5&"
            "TARGET=HTTPS%3A%2F%2Fsafe.thomson.com%2Flogin%2Fsso%2FSSOService%3Fapp%3Derrorgul_WLNQED&"
            "SMAUTHREASON=0&USER={user}&PASSWORD={password}&CAPTCHAANSWER="
        ),
        'error_gui_url': 'http://cobalttools.qed.int.westgroup.com/ErrorGUINext'
    }
}

QUERY_STRING = (
    "/initialResults.do?search=y&param=&qS=&a=&b=&c=&d=&e=&f=&g=&h=&i=&j=&x=&y=&z=&v="
    "&environs={env}&view=summary&site=all&start_time={start_time}&end_time={end_time}&submitForm=Submit"
)

TIME_FORMAT = '%Y-%m-%d %H:%M'

class ErrorReportRetriever(object):
    def __init__(self, env, user, password):
        self.env = env
        self.user = user
        self.password = password

    def get_html(self, start_time, end_time):
        session = requests.session()
        adapter = requests.adapters.HTTPAdapter(max_retries=20)
        session.mount('https://', adapter)
        session.mount('http://', adapter)

        env_data = ENV[self.env]
        headers = env_data['headers']
        headers['User-Agent'] = random.choice(USER_AGENTS)

        # SAFE ID
        payload = env_data['payload'].format(user=self.user, password=url_quote(self.password))
        session.post(env_data['safe_url'], headers=headers, data=payload, allow_redirects=True)

        # ErrorGUI request returns re-auth form expecting auto submit by JavaScript code
        error_gui_url = self._get_error_gui_url(start_time, end_time)
        auth_response = session.get(error_gui_url, headers={ 'User-Agent': random.choice(USER_AGENTS) }).text

        auth_html = pq(auth_response)
        url = auth_html('form').attr('action')
        uid = auth_html('input[name="uid"]').val()
        time = auth_html('input[name="time"]').val()
        digest = auth_html('input[name="digest"]').val()
        payload='uid={uid}&time={time}&digest={digest}'.format(uid=uid, time=url_quote(time), digest=digest)
        session.post(url, headers=headers, data=payload, allow_redirects=True)

        # get actual errors report
        report_html = session.get(error_gui_url, headers={ 'User-Agent': random.choice(USER_AGENTS) }).text

        session.close()
        
        return report_html

    def _get_error_gui_url(self, start_time, end_time):
        start = start_time.strftime(TIME_FORMAT)
        end = end_time.strftime(TIME_FORMAT)

        url = ENV[self.env]['error_gui_url'] + QUERY_STRING.format(env=self.env, start_time=url_quote(start), end_time=url_quote(end))

        return url


errorguimonitor/reporter.py
#!/usr/bin/env python

"""
Reports module for different kind of reports.
"""
import argparse
import numpy as np
import smtplib

from datetime import datetime, timedelta
from email.message import EmailMessage

from prettytable import PrettyTable

from . import __version__
from .report_builder import ErrorReportBuilder

WORKING_DAYS_CURRENT_COUNT = 1
WORKING_DAYS_PRIOR_COUNT = 5
DATE_FORMAT = '%Y-%m-%d'
RECIPIENTS = ['ihar.malkevich@thomsonreuters.com']

class ErrorCompareReporter(object):
    """
    Error compare reporter - compares periods and reports new errors or error rates increase.
    """
    def __init__(self, env, date, safeid_user, safeid_password, smtp_user, smtp_password):
        self.env = env
        self.today = date
        self.safeid_user = safeid_user
        self.safeid_password = safeid_password
        self.smtp_user = smtp_user
        self.smtp_password = smtp_password

    def run(self):
        target_dates, history_dates = self._get_report_dates(WORKING_DAYS_CURRENT_COUNT, WORKING_DAYS_PRIOR_COUNT)
        
        report_builder = ErrorReportBuilder(self.env, self.safeid_user, self.safeid_password)

        target_report = self._get_report(report_builder, target_dates)
        history_report = self._get_report(report_builder, history_dates)

        (new_errors, errors_increased) = self._prepare_report(target_report, history_report)

        (subject, body) = self._create_notification(target_dates, history_dates, new_errors, errors_increased)

        self._send_notification(subject, body, self.smtp_user, self.smtp_password, RECIPIENTS)

    def _check_error_rate_increase(self, target_errors, history_errors):
        target_median_scores = map(np.median, self._get_bootstrap_samples(np.array(target_errors), 100))
        history_median_scores = map(np.median, self._get_bootstrap_samples(np.array(history_errors), 100))

        delta_median = [x[0] - x[1] for x in zip(target_median_scores, history_median_scores)]
        conf_int_diff = self._stat_intervals(delta_median, 0.05)

        return conf_int_diff

    def _create_notification(self, target_dates, history_dates, new_errors, errors_increased):
        dates_to = '|'.join([date[0].strftime(DATE_FORMAT) for date in target_dates])
        dates_compare = '|'.join([date[0].strftime(DATE_FORMAT) for date in history_dates])

        subject = 'Error report for Website for {} v. {}'.format(dates_to, dates_compare)

        body = ''

        tbl_attr = { 
            'style': 'border: 1px solid black; border-collapse: collapse;',
            'cellpadding': 5,
            'border': 1
        }
        if len(new_errors):
            body += 'New errors observed:<br/>'
            tbl = PrettyTable()
            tbl.field_names = ['Error key', 'Occurrences']
            for key in new_errors:
                tbl.add_row([key, ', '.join([str(val) for val in new_errors[key]])])

            tbl.align["Error key"] = "l"
            body += tbl.get_html_string(attributes=tbl_attr)
        else:
            body += 'No new errors observed.<br/>'

        body += '<br/><br/>'

        if len(errors_increased):
            body += 'Errors increased:<br/>'
            tbl = PrettyTable()
            tbl.field_names = ['Error key', 'Increase rate*']
            for key in errors_increased:
                tbl.add_row([key, ' - '.join([str(round(val, 0)) for val in errors_increased[key]])])

            tbl.align["Error key"] = "l"
            body += tbl.get_html_string(attributes=tbl_attr)

            body += '<br/>* 95% confidence interval for the difference between medians'
        else:
            body += 'No errors increased observed.<br/>'

        body += '<br/><br/>Thanks'

        return (subject, body)

    def _get_bootstrap_samples(self, data, n_samples):
        indices = np.random.randint(0, len(data), (n_samples, len(data)))
        samples = data[indices]
        return samples

    def _get_report(self, report_builder, dates):
        errors = dict()
        for (start_date, end_date) in dates:
            report = report_builder.build_report(start_date, end_date)
            for module in report:
                if not module in errors:
                    errors[module] = {}

                for error in report[module]:
                    if error in errors[module]:
                        errors[module][error].append(report[module][error])
                    else:
                        errors[module][error] = [report[module][error]]

        return errors

    def _get_report_dates(self, target_days_count, history_dates_count):
        report_dates = list()

        days_collected = 0
        days_look_back = 0

        days_count = target_days_count + history_dates_count

        while days_collected < days_count:
            report_date = self.today + timedelta(days=days_look_back)
            adjust_date = False

            if report_date.weekday() == 6:
                days_look_back -= 2
                adjust_date = True
            elif report_date.weekday() == 5:
                days_look_back -= 1
                adjust_date = True

            if adjust_date == True:
                report_date = self.today + timedelta(days=days_look_back)

            start_date = datetime(report_date.year, report_date.month, report_date.day)
            end_date = datetime(report_date.year, report_date.month, report_date.day) + timedelta(hours=23, minutes=59, seconds=59)

            report_dates.append((start_date, end_date))
            
            days_collected += 1
            days_look_back -= 1

        target_dates = report_dates[:WORKING_DAYS_CURRENT_COUNT]
        history_dates = report_dates[WORKING_DAYS_CURRENT_COUNT:]

        return (target_dates, history_dates)

    def _prepare_report(self, target_report, history_report):
        errors_target = target_report['Website']
        errors_history = history_report['Website']

        new_errors = {}
        errors_increased = {}

        for key in errors_target:
            if not key in errors_history:
                new_errors[key] = errors_target[key]
            else:
                target_errors_count = errors_target[key] 
                history_errors_count = errors_history[key]

                conf_int_diff = self._check_error_rate_increase(target_errors_count, history_errors_count)
                if conf_int_diff[0] > 0:
                    errors_increased[key] = conf_int_diff

        return (new_errors, errors_increased)

    def _send_notification(self, subject, body, smtp_user, smtp_password, to, cc=None, bcc=None):
        msg = EmailMessage()

        msg['Subject'] = subject
        msg.set_content(body)
        msg['From'] = 'errorguimonitor@test.com'
        msg['To'] = ', '.join(to)
        msg.replace_header('Content-type', 'text/html')
        if cc:
            msg['Cc'] = ', '.join(cc)
        if bcc:
            msg['Bcc'] = ', '.join(bcc)

        server = smtplib.SMTP('smtp.gmail.com', 587)
        server.ehlo()
        server.starttls()
        server.login(smtp_user, smtp_password)
        server.send_message(msg)
        server.quit()

    def _stat_intervals(self, stat, alpha):
        boundaries = np.percentile(stat, [100 * alpha / 2., 100 * (1 - alpha / 2.)])
        return boundaries

def get_parser():
    parser = argparse.ArgumentParser(description='compare errors for two periods in s specific environment via the command line')

    parser.add_argument('-e', '--environment', help='the environment (QED only at the moment)', type=str)
    parser.add_argument('-d', '--date', help='the target date (year-month-day)', type=str)
    parser.add_argument('-u', '--user', help='SAFE ID user', type=str)
    parser.add_argument('-p', '--password', help='SAFE ID password', type=str)
    parser.add_argument('-smtp_user', '--smtp_user', help='Gmail account username', type=str)
    parser.add_argument('-smtp_password', '--smtp_password', help='Gmail account password', type=str)

    parser.add_argument('-v', '--version', help='displays the current version of errorguimonitor',
                        action='store_true')

    return parser

def command_line_runner():
    parser = get_parser()
    args = vars(parser.parse_args())

    if args['version']:
        print(__version__)
        return

    if not args['environment'] \
        or not args['date']:
        parser.print_help()
        return

    env = args['environment']
    try:
        target_date = datetime.strptime(args['date'], DATE_FORMAT)
    except ValueError:
        print('incorrect date provided. Please use year-month-day format')
        return

    safe_id_user = args['user']
    safe_id_password = args['password']
    smtp_user = args['smtp_user']
    smtp_password = args['smtp_password']

    if not safe_id_user \
        or not safe_id_password \
        or not smtp_user \
        or not smtp_password:
        print('Please provide usernames and passwords to successfully run the utility.')
        parser.print_help()
        return
    
    reporter = ErrorCompareReporter(env, target_date, safe_id_user, safe_id_password, smtp_user, smtp_password)
    reporter.run()

if __name__ == '__main__':
    command_line_runner()

